### ISSUE TITLE
Stage-B adds -glue-job to the --JOB-NAME while GlueJob deployer does not

### ISSUE THREAD

**Issue by dverdonschot** [awslabs/aws-serverless-data-lake-framework#312]
**Describe the bug**
Deployed SDLF using 'Hitting Production' and added the Gluejob deployer as described.

The problem that I face is that the Gluejob is made as sdlf-team-gluejobname.
But the Lambda 'sdlf-team-datasetname-routing-B' adds the jobsname as sdlf-team-gluejobname-glue-job:

SDLF-StageB > Lambda > stage-b-fetch-metadata/src/lambda_function.py line 19:
```python
        "--JOB_NAME": f"sdlf-{team}-{dataset}-glue-job",
```

Tried to fix it by changing the line to:

```python
        "--JOB_NAME": f"sdlf-{team}-{dataset}",
```
Didn't seem to help.

error:

```json
{
  "AllocatedCapacity": 12,
  "Arguments": {
    "--SOURCE_LOCATION": "s3:/s3-bucket/pre-stage/pca/legislators",
    "--job-bookmark-option": "job-bookmark-enable",
    "--OUTPUT_LOCATION": "s3:/s3-bucket-2/post-stage/pca/legislators",
    "--enable-auto-scaling": "true",
    "--JOB_NAME": "sdlf-team-legislators-glue-job"
  },
  "Attempt": 0,
  "CompletedOn": 1712227849826,
  "DPUSeconds": 0,
  "ErrorMessage": "JobName:sdlf-team-legislators and JobRunId:jr_f4bf675ca65cf9e871cfc0f98dfa2895638389cd1be5605f09df79d2b169163e failed to execute with exception Role arn:aws:iam::123456789012:role/service-role/sdlf-team-gluejobs-dev-rGlueRole-A3vhbqPFdPwJ should be given assume role permissions for Glue Service.\n (Service: AWSGlueJobExecutor; Status Code: 400; Error Code: InvalidInputException; Request ID: a7285747-6cbe-4c25-96d1-40f0cac13f7d; Proxy: null)",
  "ExecutionTime": 0,
  "GlueVersion": "4.0",
  "Id": "jr_f4bf675ca65cf9e871cfc0f98dfa2895638389cd1be5605f09df79d2b169163e",
  "JobName": "sdlf-team-legislators",
  "JobRunState": "FAILED",
  "LastModifiedOn": 1712227849826,
  "LogGroupName": "/aws-glue/jobs/sdlf-pca-glue-security-config-sdlf-pca-gluejobs-dev-rGlueRole-A3vhbqPFdPwJ",
  "MaxCapacity": 12,
  "NumberOfWorkers": 12,
  "PredecessorRuns": [],
  "SecurityConfiguration": "sdlf-team-glue-security-config",
  "StartedOn": 1712227846693,
  "Timeout": 2880,
  "WorkerType": "G.1X"
}
```

**To Reproduce**
Deploy SDLF with Gluejob deployer.
Add gluejob to sdlf-main-<domain>-<team> in transforms folder.
Run stage B, it will fail on the process data stage:

**Expected behavior**
Correct Gluejob should be specified by the fetch-metadata lambda.

**SDLF release (if known):**
E.g. 2.2.0

**Comment by cnfait** [comment]
To be frank I think `--JOB_NAME` should not be set. This is now documented as an "internal Glue argument" and the documentation explicitly says so: 
https://docs.aws.amazon.com/glue/latest/dg/aws-glue-programming-etl-glue-arguments.html
> --JOB_NAME — Internal to AWS Glue. Do not set.

I'll remove it, run a few tests and have a better look at the role issues you've mentioned.

**Comment by cnfait** [comment]
I've been able to reproduce the bug. I've prepared a fix that I'm currently testing - it should be merged in in the next couple hours.

**Comment by cnfait** [comment]
I've merged in a first update to eventually fix this issue: #315

**Comment by dverdonschot** [comment]
I tried the merged code, but still had a permission issue:

```
Resource handler returned message: "Encountered a permissions error performing a tagging operation, please add required tag permissions. See https://repost.aws/knowledge-center/cloudformation-tagging-permission-error for how to resolve. Resource handler returned message: "User: arn:aws:sts::accountid:assumed-role/sdlf-cicd-team-team/AWSCloudFormation is not authorized to perform: iam:CreateRole on resource: arn:aws:iam::accountid:role/service-role/sdlf-team-gluejobs-dev-rGlueRole-ZWcvsKkAjhpX because no identity-based policy allows the iam:CreateRole action (Service: Iam, Status Code: 403)"
```

Fixed it by adding this to sdlf-cicd/template-cicd-domain-team-role.yaml rTeamCloudFormationIamPolicy on line 426:

```yaml
              - !Sub arn:${AWS::Partition}:iam::${AWS::AccountId}:role/service-role/sdlf-${pTeamName}-* #< already existed
              - !Sub arn:${AWS::Partition}:iam::${AWS::AccountId}:role/service-role/sdlf-${pTeamName}-* #< added
```

**Comment by dverdonschot** [comment]
Now that the permissions worked, I hit a problem with the Glue Connection, where the AvailibilityZone is currently hard-coded:

```yaml
  rGlueConnection:
    Properties:
      CatalogId: !Ref AWS::AccountId
      ConnectionInput:
        ConnectionProperties: {}
        ConnectionType: NETWORK
        Description: "Network connected to the VPC data source"
        Name: !Sub sdlf-${pTeamName}-glue-conn
        PhysicalConnectionRequirements:
          AvailabilityZone: us-east-1a # this is not taking into account at all, but the property needs to be present anyway
```

Which results in error:

```
JobName:sdlf-team-legislators and JobRunId:jr_4ea9c23e577346ee8530c3ba84210b6fd9fd50e7c9 failed to execute with exception Availability Zone us-east-1a does not correspond to subnet (Service: AWSGlueJobExecutor; Status Code: 400; Error Code: InvalidInputException; Request ID: 1e56b9d1-6f0c-404a-b9e8-3530c2f9b977; Proxy: **null)**
```

Trying to fix it like this:

```yaml
          AvailabilityZone: ${AWS::Region}
```

**Comment by dverdonschot** [comment]
Now that the connection worked, there are still a few permission things that need to be fixed, when using the Glue Crawler in a Cross-account setup.

1. Bucket Policy Artifact bucket currently doesn't support Glue, fixed it by copying the "CrossAccount Team Cfn" bucket policy and making a "CrossAccount Glue Role Cfn" bucket policy.

sdlf-cicd, file template-cicd-prerequisites line 443:   rArtifactsBucketPolicy

Added:
```yaml
          - Sid: CrossAccount Glue Cfn
            Effect: Allow
            Principal:
              AWS: "*"
            Action:
              - s3:GetObject*
              - s3:GetBucket*
              - s3:List*
              - s3:DeleteObject*
              - s3:PutObject*
              - s3:Abort*
            Resource:
              - !GetAtt rArtifactsBucket.Arn
              - !Sub ${rArtifactsBucket.Arn}/*
            Condition:
              "ArnLike": !If
                - GovCloudPartition
                - "aws:PrincipalArn": !Split
                  - ","
                  - !Sub
                    - "arn:${AWS::Partition}:iam::${inner}:role/service-role/sdlf-*"
                    - inner: !Join
                      - ":role/service-role/sdlf-*,arn:aws-us-gov:iam::"
                      - Ref: "pDomainAccounts"
                - "aws:PrincipalArn": !Split
                  - ","
                  - !Sub
                    - "arn:${AWS::Partition}:iam::${inner}:role/service-role/sdlf-*"
                    - inner: !Join
                      - ":role/service-role/sdlf-*,arn:aws:iam::"
                      - Ref: "pDomainAccounts"
```

2. KMS Key Policy on sdlf-cicd-kms-key in the Devops (code) account, needs to allow GlueJob Role to decrypt the key to get the artifacts. 
  rKMSKey 

added:
```yaml
          - Sid: CrossAccount Glue Role Cfn
            Effect: Allow
            Principal: 
              - "*"
            Action:
              - kms:Decrypt
              - kms:Describe*
              - kms:Encrypt
              - kms:GenerateDataKey*
              - kms:List*
              - kms:ReEncrypt*
            Resource: "*"
            Condition:
              "ArnLike": !If
                - GovCloudPartition
                - "aws:PrincipalArn": !Split
                  - ","
                  - !Sub
                    - "arn:${AWS::Partition}:iam::${inner}:role/service-role/sdlf-*"
                    - inner: !Join
                      - ":role/service-role/sdlf-*,arn:aws-us-gov:iam::"
                      - Ref: "pDomainAccounts"
                - "aws:PrincipalArn": !Split
                  - ","
                  - !Sub
                    - "arn:${AWS::Partition}:iam::${inner}:role/service-role/sdlf-*"
                    - inner: !Join
                      - ":role/service-role/sdlf-*,arn:aws:iam::"
                      - Ref: "pDomainAccounts"
```

3. GlueJob role also needs permissions on KMS Key sdlf-cicd-kms-key in Devops account  in the Gluejobs Role:

Currently I don't see a way to do this in the code, as there is no SSM key for this KMS Key in the Producer account that I can link to. For now I added it manually to our role for the Gluejob deployer.

**Comment by cnfait** [comment]
Hey @dverdonschot - sorry, I need to communicate better on this issue: I do not consider it fixed yet, and there is more work coming to make it more robust. I'm travelling/on customer workshops this week so this make it hard for me to fix the issue in a timely fashion but work is definitely happening on this. 

Just to share the work in progress:
ae96f96140fa6d765241333c5d4bb2192be2322e
In this commit I'm doing two things:
- using a file hash instead of the git reference as suffix for the glue jobs pushed to S3. This is to avoid triggering a Glue job update when they haven't changed: https://github.com/awslabs/aws-serverless-data-lake-framework/commit/ae96f96140fa6d765241333c5d4bb2192be2322e#diff-77a1b40afa1c17f634e30de489542a8c51a6ab9135289da6815b165e32aa82a4R196-R201
- update S3, KMS resource policies (child account - sdlf-foundations) and IAM policies (devops account - sdlf-cicd). As you correctly mentioned there is no meaningful way of getting the ARNs when we need them for these. Here I'm using a few things to still keep things secure: conditions, account ids - I'll let you check the commit for the exact details.

The last missing piece is for the CodeBuild job to be able to retrieve the S3 bucket name at runtime. Current thinking is to share a SSM parameter from the child account to the devops account (d5fcd0e2ca355ff4d2897f5144f4b1a3f9204fda).

I know all this is bigger than anticipated - in the middle of investigating the issue I've been thinking using the child account's artifacts bucket would be better in this case (as Glue jobs references S3 directly and the script is loaded dynamically from S3 when edited in the console) and as is often the case when taking that kind of decisions this is taking a bit longer than planned!

**Comment by cnfait** [comment]
I believe this is now entirely covered by PR #316.

If you wish to update your environment I would roughly follow these steps:
- make sure you have an up to date clone of the GitHub SDLF repository
- from inside that clone, rerun `./deploy.sh crossaccount-cicd-roles` for all domain accounts
- rerun `./deploy.sh devops-account`

`./deploy.sh devops-account` is going to complain about failing to create/push git branches, this can be ignored.

-  Make sure the `rMain` CodePipelines are not running. (if running, "stop execution and wait" in the web console)
- Update `template.yaml` in `sdlf-foundations` to be up to date
- Update `template.yaml` in `sdlf-team` to be up to date
- Update the content of the `sdlf-cicd` repository (on each branch you want to test the changes on) on CodeCommit to be up to date
- `rMain` CodePipeline should be triggered automatically after updating `sdlf-cicd`, wait for it to finish
- `rDomain` CodePipeline should be triggered automatically after `rMain` finishes, wait for it to finish

**Comment by cnfait** [comment]
I've created a separate issue about the Glue Connection: https://github.com/awslabs/aws-serverless-data-lake-framework/issues/317

**Comment by dverdonschot** [comment]
I tried rolling out a new environment based on the main branch.
But I am hitting a problem on the PublishDomainDetailsInDevops stage in the domain pipeline.
It basicly fails because it cannot find the variable oS3ArtifactsBucket created by the previous "CreateStack" step.

```yaml              
  "pArtifactsBucket" : "#{DomainVariables.oS3ArtifactsBucket}",
```

One thing I do notice is the value is called <domain>oS3ArtifactsBucket, for example rRdpcaoS3ArtifactsBucket, not just oS3ArtifactsBucket. But when I adjusted for that value it still failed.

In the end I just put the bucketname hardcoded in the value (we only have 1 domain anyway).
That solved the problem and now the SSM parameter is made.

**Comment by cnfait** [comment]
This is something I need to update documentation-wise. I'll close the ticket for now - a big documentation update is coming "soon" (this month) so that should help.

I'll publish a new version of SDLF (2.3.0) incorporating these changes and a few others. The next version after that should be published early May - I want the cycle to be faster for a while as we're getting a lot of feedback and several people inside AWS are working on much-asked features.

**Comment by cnfait** [comment]
I've finally released [SDLF 2.3.0](https://github.com/awslabs/aws-serverless-data-lake-framework/releases/tag/2.3.0) which should solve #312 and #317 (with a simpler solution in the end).
